### PR TITLE
Reindex SKU(s) Feature 

### PR DESCRIPTION
**Pull request by bsuravech**
**Summary**
Backport the M2 feature for Reindexing Product SKUs in admin. Changes to match:
https://github.com/algolia/algoliasearch-magento-2/pull/536

**Result**
+ Adding new admin menu section for Algolia Search under System to include quick link to Algolia's system configuration section and to the reindex SKU(s) feature.  
+ Adminhtml controller directory Algoliasearch has been added for proper url structure and in result updated queue controller links for notifications. 
+ ACL has been added for Reindex SKU feature.
+ Created new exception classes to match M2 with update to UnknownSkuException to ProductUnknownSkuException naming. 
+ Updated reindexPostAction logic to load collection instead of loading products individually as best practice for optimisation. 
+ Updated parameter $skus filter to preg_split array_map('trim') array_filter (to knock out spacing and blank entries)
+ Updated ProductNotVisibleException catch to not retry parent product in the catch handle but on the checkAndReindex() method to retry there (this is to prevent Exception Inception where if the parent product threw an error it won't throw a fatal error within the Exception catch)
+ Added and updated canProductBeReindexed method in data helper to match M1 products requirements.

### CODE REVIEW

File: app/code/community/Algolia/Algoliasearch/Helper/Data.php [hunk 1]
```diff
@@ -437,6 +437,46 @@ public function rebuildStoreCategoryIndexPage($storeId, $collectionDefault, $pag
...
+    /**
+     * Check if product can be index on Algolia
+     *
+     * @param Mage_Catalog_Model_Product $product
+     * @param int     $storeId
+     *
+     * @return bool
+     *
+     */
+    public function canProductBeReindexed(Mage_Catalog_Model_Product $product, $storeId)
```
Comment (JanPetr) [inline app/code/community/Algolia/Algoliasearch/Helper/Data.php]: How about moving this method to ProductHelper? It's very Product specific method.

File: app/code/community/Algolia/Algoliasearch/Helper/Data.php [hunk 2]
```diff
@@ -437,6 +437,46 @@ public function rebuildStoreCategoryIndexPage($storeId, $collectionDefault, $pag
...
+    {
+        if ($product->isDeleted() === true) {
+            throw (new Algolia_Algoliasearch_Model_Exception_ProductDeletedException())
+                ->withProduct($product)
+                ->withStoreId($storeId);
+        }
+
+        if ($product->getStatus() == Mage_Catalog_Model_Product_Status::STATUS_DISABLED) {
```
Comment (JanPetr) [inline app/code/community/Algolia/Algoliasearch/Helper/Data.php]: Can a strict (`===`) comparison be used?

File: app/code/community/Algolia/Algoliasearch/Helper/Entity/Producthelper.php [hunk 3]
```diff
@@ -1041,6 +1043,44 @@ public function getObject(Mage_Catalog_Model_Product $product)
...
+     * @param array $productIds
+     * @return array
+     */
+    public function getParentProductIds(array $productIds)
+    {
+        $parentIds = array();
+        foreach ($this->getCompositeTypes() as $typeInstance) {
+            $parentIds = array_merge($parentIds, $typeInstance->getParentIdsByChild($productIds));
+        }
+        return $parentIds;
```
Comment (JanPetr) [inline app/code/community/Algolia/Algoliasearch/Helper/Entity/Producthelper.php]: Can you leave a blank line before the `return`? It's a bit more readable :)

File: app/code/community/Algolia/Algoliasearch/Helper/Entity/Producthelper.php [hunk 4]
```diff
@@ -1041,6 +1043,44 @@ public function getObject(Mage_Catalog_Model_Product $product)
...
+     * @return Mage_Catalog_Model_Product_Type[]
+     *
+     * @see Mage_Catalog_Model_Resource_Product_Flat_Indexer::getProductTypeInstances()
+     */
+    private function getCompositeTypes()
+    {
+        if ($this->compositeTypes === null) {
+            /** @var Mage_Catalog_Model_Product $productEmulator */
+            $productEmulator = Mage::getModel('catalog/product');
+            /** @var Mage_Catalog_Model_Product_Type $productType */
```
Comment (JanPetr) [inline app/code/community/Algolia/Algoliasearch/Helper/Entity/Producthelper.php]: Can you leave a blank line between the `$productEmulator = ...` and `/** @var ...`? :)

File: app/code/community/Algolia/Algoliasearch/Helper/Data.php (hunk 1 above)
Comment (bsuravech) [reply app/code/community/Algolia/Algoliasearch/Helper/Data.php]: We could move that over there! We should consider that as well for M2 feature if we want to keep it the same.

File: app/code/community/Algolia/Algoliasearch/Helper/Data.php (hunk 2 above)
Comment (bsuravech) [reply app/code/community/Algolia/Algoliasearch/Helper/Data.php]: Strict typing threw issues on travis interestingly enough. $product->getStatus() will not guarantee a int return to match strict typing.

File: app/code/community/Algolia/Algoliasearch/Helper/Entity/Producthelper.php (hunk 3 above)
Comment (bsuravech) [reply app/code/community/Algolia/Algoliasearch/Helper/Entity/Producthelper.php]: You got it!

File: app/code/community/Algolia/Algoliasearch/Helper/Entity/Producthelper.php (hunk 4 above)
Comment (bsuravech) [reply app/code/community/Algolia/Algoliasearch/Helper/Entity/Producthelper.php]: Will add spacing there!

File: app/code/community/Algolia/Algoliasearch/Helper/Data.php (hunk 2 above)
Comment (JanPetr) [reply app/code/community/Algolia/Algoliasearch/Helper/Data.php]: Magento /o\ OK :))

**Pull request merged**